core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala

```diff
@@ -284,18 +310,52 @@ private[ui] class StagePagedTable(
           {UIUtils.makeProgressBar(started = stageData.numActiveTasks,
           completed = stageData.numCompleteTasks, failed = stageData.numFailedTasks,
           skipped = 0, reasonToNumKilled = stageData.killedTasksSummary, total = info.numTasks)}
-        </td>
-        <td>{data.inputReadWithUnit}</td>
-        <td>{data.outputWriteWithUnit}</td>
-        <td>{data.shuffleReadWithUnit}</td>
-        <td>{data.shuffleWriteWithUnit}</td> ++
-        {
-          if (isFailedStage) {
-            failureReasonHtml(info)
-          } else {
-            Seq.empty
-          }
-        }
+        </td> ++
+        {if (metrics.contains("Input")) {
+          <td>{data.inputReadWithUnit}</td>
+        } else {
+          Seq.empty
+        }} ++
+        {if (metrics.contains("Output")) {
```

Comment:
Could we find a way to avoid all this duplicate code? Something like iterate over optionalHeaders.keys or something like that.

Reply:
fixed